### PR TITLE
Add visualDensity and focus support to ListTile

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

This adds `visualDensity` and focus support to the `ListTile` widget.

It actually sort of already had focus support, since it was using an `InkWell` it would already respond to focus changes, this just adds attributes for `focusNode`, `focusColor`, and `hoverColor` so that they may be passed through to the `InkWell`.

For `visualDensity` support, in addition to allowing visual density to affect the height of the tile, I added some control over spacing vertically between the subtitle and the title widgets.  Horizontally, it controls the size of the gaps between the leading and trailing widgets and the title and subtitle widgets. The width of the tile isn't affected, since it expands to fill its parent's width.

This PR doesn't affect the layout of `ListTile` with the default `visualDensity` of 0.0.

## Related Issues

- https://github.com/flutter/flutter/issues/43350

## Tests

- Added tests for focus, hover, and density controls on `ListTile`.

## Breaking Change

- [X] No, this is *not* a breaking change.

### CODE REVIEW

File: dev/manual_tests/lib/density.dart [hunk 1]
```diff
@@ -252,7 +252,8 @@ class _OptionsState extends State<Options> {
                           min: VisualDensity.minimumDensity,
                           max: VisualDensity.maximumDensity,
                           onChanged: (double value) {
-                            widget.model.density = widget.model.density.copyWith(horizontal: value, vertical: widget.model.density.vertical);
+                            widget.model.density = widget.model.density.copyWith(
+                                horizontal: value, vertical: widget.model.density.vertical);
```
Comment (goderbauer) [inline dev/manual_tests/lib/density.dart:256]: nit: the formatting seems off. I would suggest the following

```suggestion
                            widget.model.density = widget.model.density.copyWith(
                              horizontal: value,
                              vertical: widget.model.density.vertical,
                            );
                            
```

Comment (goderbauer) [reply dev/manual_tests/lib/density.dart:256]: Same below.

File: dev/manual_tests/lib/density.dart [hunk 2]
```diff
@@ -376,7 +378,9 @@ class _ControlTile extends StatelessWidget {
         padding: const EdgeInsets.all(8.0),
         child: Column(
           children: <Widget>[
-            Align(alignment: AlignmentDirectional.topStart, child: Text(label, textAlign: TextAlign.start)),
+            Align(
+                alignment: AlignmentDirectional.topStart,
+                child: Text(label, textAlign: TextAlign.start)),
```
Comment (goderbauer) [inline dev/manual_tests/lib/density.dart:383]: nit: format (add a trailing comma, un-indent this and previous line by 2, and put the closing ) on the next line?

File: packages/flutter/lib/src/material/list_tile.dart [hunk 3]
```diff
@@ -633,11 +634,16 @@ class ListTile extends StatelessWidget {
...
+    this.visualDensity,
     this.contentPadding,
     this.enabled = true,
     this.onTap,
     this.onLongPress,
     this.selected = false,
+    this.focusColor,
+    this.hoverColor,
+    this.focusNode,
+    this.autofocus = false,
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/list_tile.dart:646]: is autofocus == null allowed? If not, document and assert?

File: packages/flutter/test/material/list_tile_test.dart [hunk 4]
```diff
@@ -1184,4 +1188,209 @@ void main() {
...
+      );
+    }
+    await tester.pumpWidget(buildApp());
+
+    await tester.pumpAndSettle();
+    expect(focusNode.hasPrimaryFocus, isTrue);
+    expect(
+      Material.of(tester.element(find.byKey(tileKey))),
+      paints
+        ..rect(
```
Comment (goderbauer) [inline packages/flutter/test/material/list_tile_test.dart:1224]: nit: formatting here and elsewhere: use a trailing comma and put closing ) on the next line?

**Review by goderbauer — APPROVED** [review]
LGTM

File: dev/manual_tests/lib/density.dart (hunk 1 above)
Comment (gspencergoog) [reply dev/manual_tests/lib/density.dart]: Fixed, and below.

File: dev/manual_tests/lib/density.dart (hunk 2 above)
Comment (gspencergoog) [reply dev/manual_tests/lib/density.dart]: Fixed

File: packages/flutter/lib/src/material/list_tile.dart (hunk 3 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/list_tile.dart:646]: Fixed by merging to head, where another PR added autofocus.

File: packages/flutter/test/material/list_tile_test.dart (hunk 4 above)
Comment (gspencergoog) [reply packages/flutter/test/material/list_tile_test.dart:1272]: Done.

**Review by HansMuller — APPROVED** [review]
LGTM too

**Pull request merged**